File: Doc/Kattan/Solutions-Manual/problem_5_1.m
% =========================================================================
% Problem 5.1 - Plane Truss with Nine Elements (Fig. 5.5)
% Reference: P. I. Kattan, "MATLAB Guide to Finite Elements: An Interactive
% Approach" - Chapter 5, plane truss element formulation.
%
% =========================================================================
% Fig. 5.5. Plane Truss for Problem 5.1
% =========================================================================
%
%     Y
%     ^
%     |   2 ---(6)--- 4
%     |  / \       /  \
%     | /   \ (5) / (9)\
%   7m|/ (3) \  /      \
%     |1---(2)--3--(4)--5--(8)--6-----> X
%     |      5m      5m      5m
%     |       20 kN (+X at node 2)
%
%     Node coordinates:
%       1: (0, 0)    2: (5, 7)    3: (5, 0)
%       4: (10, 7)   5: (10, 0)   6: (15, 0)
%
%     Elements (9): 1→2, 1→3, 2→3, 3→5, 2→5, 2→4, 4→5, 5→6, 4→6
%
% =========================================================================
%
% Computes:
%   1. Global stiffness matrix K
%   2. Displacements at nodes 2, 3, 4, and 5
%   3. Reactions at nodes 1 and 6
%   4. Stress in each element
% =========================================================================

clear; clc;

E = 210e6

A = 0.005

L1 = PlaneTrussElementLength(0,0,5,7)

L5 = PlaneTrussElementLength(0,0,5,-7)

L9 = PlaneTrussElementLength(0,0,5,-7)

theta1 = atan(7/5)*180/pi

theta2 = 0

theta3 = 270

theta4 = 0

theta5 = 360 - theta1

theta6 = 0

theta7 = 270

theta8 = 0

theta9 = theta5

k1 = PlaneTrussElementStiffness(E,A,L1,theta1)

k2 = PlaneTrussElementStiffness(E,A,5,theta2)

k3 = PlaneTrussElementStiffness(E,A,7,theta3)

k4 = PlaneTrussElementStiffness(E,A,5,theta4)

k5 = PlaneTrussElementStiffness(E,A,L5,theta5)

k6 = PlaneTrussElementStiffness(E,A,5,theta6)

k7 = PlaneTrussElementStiffness(E,A,7,theta7)

k8 = PlaneTrussElementStiffness(E,A,5,theta8)

k9 = PlaneTrussElementStiffness(E,A,L9,theta9)

K = zeros(12,12)

K = PlaneTrussAssemble(K,k1,1,2)

K = PlaneTrussAssemble(K,k2,1,3)

K = PlaneTrussAssemble(K,k3,2,3)

K = PlaneTrussAssemble(K,k4,3,5)

K = PlaneTrussAssemble(K,k5,2,5)

K = PlaneTrussAssemble(K,k6,2,4)

K = PlaneTrussAssemble(K,k7,4,5)

K = PlaneTrussAssemble(K,k8,5,6)

K = PlaneTrussAssemble(K,k9,4,6)

k = K(3:10,3:10)

f = [20 ; 0 ; 0 ; 0 ; 0 ; 0 ; 0 ; 0]

u = k\f

U = [0 ; 0 ; u ; 0 ; 0]

F = K*U

u1 = [U(1) ; U(2) ; U(3) ; U(4)]

sigma1 = PlaneTrussElementStress(E,L1,theta1,u1)

u2 = [U(1) ; U(2) ; U(5) ; U(6)]

sigma2 = PlaneTrussElementStress(E,5,theta2,u2)

u3 = [U(3) ; U(4) ; U(5) ; U(6)]

sigma3 = PlaneTrussElementStress(E,7,theta3,u3)

u4 = [U(5) ; U(6) ; U(9) ; U(10)]

sigma4 = PlaneTrussElementStress(E,5,theta4,u4)

u5 = [U(3) ; U(4) ; U(9) ; U(10)]

sigma5 = PlaneTrussElementStress(E,L5,theta5,u5)

u6 = [U(3) ; U(4) ; U(7) ; U(8)]

sigma6 = PlaneTrussElementStress(E,5,theta6,u6)

u7 = [U(7) ; U(8) ; U(9) ; U(10)]

sigma7 = PlaneTrussElementStress(E,7,theta7,u7)

u8 = [U(9) ; U(10) ; U(11) ; U(12)]

sigma8 = PlaneTrussElementStress(E,5,theta8,u8)

u9 = [U(7) ; U(8) ; U(11) ; U(12)]

sigma9 = PlaneTrussElementStress(E,L9,theta9,u9)

File: problems/problem_5_1.jl
#!/usr/bin/env julia
# ═══════════════════════════════════════════════════════════════════════
# Problem 5.1 — Plane Truss (Fig 5.5)
# Reference: P. I. Kattan, "MATLAB Guide to Finite Elements:
#   An Interactive Approach" (2nd ed., Springer, 2007)
# ═══════════════════════════════════════════════════════════════════════
#
#                           2           4
#               20 kN       O-----------O
#               --------->/ | \         | \
#                       /   |   \       |   \
#                     /     |     \     |     \
#                   /       |       \   |       \
#                 /         |         \ |         \
#               /           |           \           \
#               O-----------O-----------O-----------O
#               1           3           5           6
#              / \                                 / \
#             ////                                ////
#
#               |<-- 5 m -->|<-- 5 m -->|<-- 5 m -->|
# ═══════════════════════════════════════════════════════════════════════
# Computes:
#   1. Global stiffness matrix K (12×12)
#   2. Displacements at nodes 2-5
#   3. Reactions at nodes 1 and 6
#   4. Stress in each of the 9 elements
# ═══════════════════════════════════════════════════════════════════════

using LibFEM
using LinearAlgebra

# ─── Parameters ──────────────────────────────────────────────
E = 210e6
A = 0.005

# ─── Node coordinates ────────────────────────────────────────
# Node 1 (0,0) pinned, Node 6 (15,0) pinned/roller
# Nodes: 1(0,0), 2(5,7), 3(5,0), 4(10,7), 5(10,0), 6(15,0)
x = [0.0, 5.0, 5.0, 10.0, 10.0, 15.0]
y = [0.0, 7.0, 0.0, 7.0, 0.0, 0.0]

# ─── Element data ────────────────────────────────────────────
# Format: (node_i, node_j)
elements = [
    (1, 2),  # 1
    (1, 3),  # 2
    (2, 3),  # 3
    (3, 5),  # 4
    (2, 5),  # 5
    (2, 4),  # 6
    (4, 5),  # 7
    (5, 6),  # 8
    (4, 6),  # 9
]

# ─── Element stiffness matrices ──────────────────────────────
println("=== Element stiffness matrices ===")
L_vals = Float64[]
theta_vals = Float64[]
k_vals = []

for (i, (n1, n2)) in enumerate(elements)
    dx = x[n2] - x[n1]
    dy = y[n2] - y[n1]
    L = sqrt(dx^2 + dy^2)
    theta = atan(dy, dx) * 180 / pi
    push!(L_vals, L)
    push!(theta_vals, theta)
    k = d2_truss_elementstiffness(E, A, L, theta)
    push!(k_vals, k)
    println("k$i (nodes $n1-$n2, L=$L, theta=$theta) =")
    display(k)
end

# ─── Assembly ────────────────────────────────────────────────
K = zeros(12, 12)
for (idx, (n1, n2)) in enumerate(elements)
    d2_truss_assemble(K, k_vals[idx], n1, n2)  # modifies K in-place via _assemble!
end

println("\nK =")
display(K)

# ─── Solve (free DOFs = 3:10, nodes 2-5 × 2 DOF) ────────────
k = K[3:10, 3:10]
f = [20.0; 0.0; zeros(6)...]  # 20 kN at node 2 in x-direction

u = k \ f
# Full displacement: nodes 1 (1,2)=0, nodes 2-5 (3:10)=u, node 6 (11,12)=0
U = [0.0; 0.0; u; 0.0; 0.0]
F = K * U

println("\nk (reduced) =")
display(k)
println("\nf =")
display(f)
println("\nu =")
display(u)
println("\nU =")
display(U)
println("\nF =")
display(F)

# ─── Post-processing: stresses ───────────────────────────────
println("\n=== Element stresses ===")
sigmas = Float64[]
for (idx, (n1, n2)) in enumerate(elements)
    u_elem = [U[2*n1-1]; U[2*n1]; U[2*n2-1]; U[2*n2]]
    sigma = d2_truss_elementstress(E, L_vals[idx], theta_vals[idx], u_elem)
    push!(sigmas, sigma)
    println("sigma$idx (nodes $n1-$n2) = ", sigma)
end

# ─── Equilibrium check ───────────────────────────────────────
println("\n--- Equilibrium check ---")
println("Applied force: 20 kN at node 2 (x-direction)")
println("Reaction at node 1 (F1,F2): ", F[1], ", ", F[2])
println("Reaction at node 6 (F11,F12): ", F[11], ", ", F[12])
println("Sum Fx = ", sum(F[1:2:end]), " (should be 20)")
println("Sum Fy = ", sum(F[2:2:end]), " (should be 0)")

# ─── Self-validation ─────────────────────────────────────────
# Expected values verified against Octave execution of Kattan's problem_5_1.m
K_expected = [
    251236.0   57731.1   -41236.5   -57731.1  -210000.0        0.0        0.0        0.0        0.0        0.0        0.0       0.0
     57731.1   80823.5   -57731.1   -80823.5        0.0        0.0        0.0        0.0        0.0        0.0        0.0       0.0
    -41236.5  -57731.1   292473.0        0.0        0.0        0.0  -210000.0        0.0   -41236.5    57731.1        0.0       0.0
    -57731.1  -80823.5        0.0   311647.0        0.0  -150000.0        0.0        0.0    57731.1   -80823.5        0.0       0.0
   -210000.0       0.0        0.0        0.0   420000.0        0.0        0.0        0.0  -210000.0        0.0        0.0       0.0
         0.0       0.0        0.0  -150000.0        0.0   150000.0        0.0        0.0        0.0        0.0        0.0       0.0
         0.0       0.0  -210000.0        0.0        0.0        0.0   251236.0   -57731.1        0.0        0.0   -41236.5   57731.1
         0.0       0.0        0.0        0.0        0.0        0.0   -57731.1   230824.0        0.0  -150000.0    57731.1  -80823.5
         0.0       0.0   -41236.5    57731.1  -210000.0        0.0        0.0        0.0   461236.0   -57731.1  -210000.0      0.0
         0.0       0.0    57731.1   -80823.5        0.0        0.0        0.0  -150000.0   -57731.1   230824.0        0.0       0.0
         0.0       0.0        0.0        0.0        0.0        0.0   -41236.5    57731.1  -210000.0        0.0   251236.0  -57731.1
         0.0       0.0        0.0        0.0        0.0        0.0    57731.1   -80823.5        0.0        0.0   -57731.1   80823.5
]
@assert isapprox(K, K_expected; rtol=1e-8, atol=1e-10) "K mismatch"
@assert isapprox(u, [0.00020834281842258584, -3.3338372385991406e-5, 1.0582010582010574e-5, -3.333837238599142e-5, 0.0001765967866765541, 1.0662635424540202e-5, 2.1164021164021147e-5, -5.155958679768199e-5]; rtol=1e-8) "u mismatch"
@assert isapprox(F[1], -8.888888888888882; rtol=1e-10) "F[1] mismatch"
@assert isapprox(F[2], -9.333333333333329; rtol=1e-10) "F[2] mismatch"
@assert isapprox(F[3], 20.0; rtol=1e-10) "F[3] mismatch"
@assert isapprox(F[11], -11.1111111111111; rtol=1e-10) "F[11] mismatch"
@assert isapprox(F[12], 9.333333333333323; rtol=1e-10) "F[12] mismatch"
@assert isapprox(sigmas[1], 2293.953404544699; rtol=1e-8) "sigma1 mismatch"
@assert isapprox(sigmas[5], -2293.953404544699; rtol=1e-8) "sigma5 mismatch"
@assert isapprox(sigmas[2], 444.4444444444441; rtol=1e-8) "sigma2 mismatch"
@assert isapprox(sigmas[3], 0.0; rtol=1e-8, atol=1e-12) "sigma3 mismatch"
@assert isapprox(sigmas[4], 444.4444444444441; rtol=1e-8) "sigma4 mismatch"
@assert isapprox(sigmas[6], -1333.3333333333335; rtol=1e-8) "sigma6 mismatch"
@assert isapprox(sigmas[7], 1866.6666666666665; rtol=1e-8) "sigma7 mismatch"
@assert isapprox(sigmas[8], -888.8888888888882; rtol=1e-8) "sigma8 mismatch"
@assert isapprox(sigmas[9], -2293.953404544698; rtol=1e-8) "sigma9 mismatch"
